simulator/func_sim/rf/t/unit_test.cpp

```diff
@@ -0,0 +1,150 @@
+/**
+ * Unit tests for rf
+ * @author Alexander Misevich
+ * Copyright 2018 MIPT-MIPS
+ */
+
+// generic C
+#include <cassert>
+#include <cstdlib>
+#include <cmath>
+
+// Google Test library
+#include <gtest/gtest.h>
+
+// MIPS-MIPS modules
+#include <mips/mips.h>
+#include "../rf.h"
+
+#define GTEST_ASSERT_NO_DEATH(statement) \
+    ASSERT_EXIT({{ statement } ::exit(EXIT_SUCCESS); }, ::testing::ExitedWithCode(0), "")
+
+static_assert(MIPSRegister::MAX_REG >= 32);
+
+// Testing methods of the class
+TEST( Rf, empty_rf)
+{
+    // Create array of 35 Regs
+    ASSERT_FALSE( RF<MIPS>().array.empty());    
+}
+
+TEST( RF, read_write_rf)
+{
+    auto rf = std::make_unique<RF<MIPS>>();
+    
+    // Fill array using write() and check correctness using read()
+    for( size_t i = 0; i < 32; ++i)
+    {   
+        rf->write( MIPSRegister(i), i);
```

Comment:
Oh. The original intention is to have these methods in private scope, but due to some reasons they are public.

Anyway, these tests are very good and should be kept even if the methods were moved to private scope.
How is it possible? There are two ways:

1. Hacker way. Never do this at home.
```c++
#define private public
#include "../rf.h"
#undef private
```

2. More safe and elegant way is to move methods you want to test to **protected** scope and define a following class in `unit_test.cpp`:

```c++
class TestRF : public RF {
public:
    // ctors

    // list of methods you want to use from basic class
    using RF::read;
    using RF::write;
};
```

Reply:
Is it needed to put array in protected scope?

Reply:
I would keep array and get_entry inside private scope so we can switch container later if we need to. Thus, you have to remove `array.empty() == false` test.